C++ components unit test framework

jpeletier opened the pull request:
# What does this implement/fix?

![image](https://github.com/user-attachments/assets/a7e70e00-f4a5-4744-8a47-5fbe4b78a079) 
This PR introduces a simple framework to enable writing unit tests in C++ for components. Most of the components in ESPHome currently don't have unit tests that actually verify functionality and avoid regressions.

Unit tests are located in `tests/components/<component name>`

This PR also includes tests for the `uart` component to illustrate the usage and serve as actual tests for it.

# How to write C++ ESPHome unit tests

1. Locate the folder with your component or create a new one with the same name as the component.
2. Write the tests. You can add as many `.cpp` and `.h` files as you need to organize your tests.

## Running component unit tests

(from the repository root)
```bash
./script/cpp_unit_test.py component1 component2 ...
```

The above will compile and run the provided components and their tests.

To run a specific test suite, you can provide a Google Test filter:

```bash
GTEST_FILTER='UART*' ./script/cpp_unit_test.py uart modbus
```

The process will return `0` for succcess or nonzero for failure. In case of failure, the errors will be printed out to the console.

Here is an example `cpp` test file

```cpp

#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include "esphome/components/uart/uart_component.h"

class MockUARTComponent : public UARTComponent {
 public:
  std::vector<uint8_t> written_data;
  void write_array(const uint8_t *data, size_t len) override { written_data = std::vector(data, data + len); }
  MOCK_METHOD(bool, read_array, (uint8_t * data, size_t len), (override));
  MOCK_METHOD(bool, peek_byte, (uint8_t * data), (override));
  MOCK_METHOD(int, available, (), (override));
  MOCK_METHOD(void, flush, (), (override));
  MOCK_METHOD(void, check_logger_conflict, (), (override));
};

TEST(UARTComponentTest, SetGetBaudRate) {
  MockUARTComponent mock;
  mock.set_baud_rate(38400);
  EXPECT_EQ(mock.get_baud_rate(), 38400);
}

TEST(UARTComponentTest, SetGetStopBits) {
  MockUARTComponent mock;
  mock.set_stop_bits(2);
  EXPECT_EQ(mock.get_stop_bits(), 2);
}

TEST(UARTComponentTest, WriteArrayVector) {
  MockUARTComponent mock;
  std::vector<uint8_t> data = {10, 20, 30};
  mock.write_array(data);
  EXPECT_EQ(mock.written_data, data);
}

// ...

```

When running `./script/cpp_unit_test.py uart` (example where were are testing the `uart` component), you will get an output like the below:

```bash
[==========] Running 22 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 10 tests from UARTComponentTest
[ RUN      ] UARTComponentTest.SetGetBaudRate
[       OK ] UARTComponentTest.SetGetBaudRate (0 ms)
[ RUN      ] UARTComponentTest.SetGetStopBits
[       OK ] UARTComponentTest.SetGetStopBits (0 ms)
[ RUN      ] UARTComponentTest.SetGetDataBits
[       OK ] UARTComponentTest.SetGetDataBits (0 ms)
[ RUN      ] UARTComponentTest.SetGetParity
[       OK ] UARTComponentTest.SetGetParity (0 ms)
[ RUN      ] UARTComponentTest.SetGetRxBufferSize
[       OK ] UARTComponentTest.SetGetRxBufferSize (0 ms)
[ RUN      ] UARTComponentTest.WriteArrayVector
[       OK ] UARTComponentTest.WriteArrayVector (0 ms)
[ RUN      ] UARTComponentTest.WriteByte
[       OK ] UARTComponentTest.WriteByte (0 ms)
[ RUN      ] UARTComponentTest.WriteStr
[       OK ] UARTComponentTest.WriteStr (0 ms)
[ RUN      ] UARTComponentTest.ReadByteSuccess
[       OK ] UARTComponentTest.ReadByteSuccess (0 ms)
[ RUN      ] UARTComponentTest.ReadByteFailure
[       OK ] UARTComponentTest.ReadByteFailure (0 ms)
[----------] 10 tests from UARTComponentTest (0 ms total)
```

CI workflow has been modified so that it detects exactly what C++ tests actually need to run based on what code changed.

If core C++ files are changed, all C++ tests are run.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] Host
- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx


## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

bdraco commented on tests/integration/test_api_conditional_memory.py:
```diff
@@ -115,7 +115,7 @@ def on_state(state: EntityState) -> None:
             # Wait for initial states
-            await asyncio.wait_for(states_future, timeout=5.0)
+            await asyncio.wait_for(states_future, timeout=60)
 
```
Please revert this change. The race should already be fixed on dev

jpeletier replied:
Done

bdraco commented on tests/unit_tests/components/README.md:
```diff
@@ -0,0 +1,26 @@
+# How to write C++ ESPHome unit tests
+
+1. Copy the `example` folder as a starting template and rename it with your component name.
```
I think we likely want these to run with the component_tests

jpeletier replied:
But aren't component_tests more like code generation tests? I think that'd be a separate feature. In here we're unit testing a component's cpp code.

bdraco replied:
The key issue is all the python tests run in the same job runner so when we start adding more tests we need to make sure we are only running the correct tests for each component.

It might make more sense to make a new directory for these type of tests.

jpeletier replied:
I moved the c++ tests to sit with `tests/components` which already had a folder per component.
I thought about a new directory but perhaps it would be confusing to have yet another directory that has a folder per component.

So now `tests/components` will have a mix of yaml and .cpp/.h with unit tests.

bdraco commented on tests/unit_tests/components/test001.yaml:
```diff
@@ -0,0 +1,24 @@
+  test_common: !include common.yaml
+
+esphome:
+  # Make sure the configuration name matches the file name so they are unique:
```
Ideally this is all automatic based on where the test files are.  We likely don't want to maintain tons of these boilerplate configs.

Can we inject all this like the integrations tests do?

jpeletier replied:
These would not be a file per component, but rather a config that has as many components as possible that are compatible with each other. This is so that the tests run fast: we compile a big config and run all the tests. Note that it does not matter if the configuration makes sense since it is not really run in any way. It is just there so that the C++ code is pulled into the build. Components are instantiated manually in each test.

Alternatively, if it was possible to just indicate in the config that a component's C++ code must be available (copied to the build folder), this file wouldn't really be needed. I haven't seen a way to do that, please let me know if there is something that could do the trick and I am missing.

> Can we inject all this like the integrations tests do?

I am not sure what you mean. Can you point me to an example?

bdraco replied:
Thanks for the color.  I'm not sure this design is going to scale well if 100+ components have tests.  While that will be a good problem to have, we should think about a design now that will handle that.

jpeletier replied:
ok so I gave this a complete turnaround.

1. I wrote a new script `script/cpp_unit_test.py`, it takes component names separated by spaces. It will build all tests for those components into a single platformio project, then runs them. It may also take the special switch `--all` in which case it compiles and runs all unit tests available in ESPHome. (currently only `uart`).
2. I modified the new `determine-jobs.py`. It now also outputs what c++ unit tests to run, if any. C++ unit tests will run when any of the following conditions are met:
    * Any C++ core source files changed (esphome/core/*), in which case all cpp unit tests run.
    * A test file for a component changed, which triggers tests for that component only.
    * The code for a component changed, which triggers tests for that component and all components that depend on it.

3. Added a new CI job, C++ Unit tests, which is only launched if there are actually any tests to run.

bdraco commented on script/determine-jobs.py:
```diff
@@ -50,6 +50,29 @@
 
+def core_changed(files: list[str]) -> bool:
+    """Check if any core C++ or Python files have changed."""
+    for file in files:
+        if file.startswith("esphome/core/") and file.endswith(CPP_FILE_EXTENSIONS):
+            return True
+        if file.startswith("esphome/core/") and file.endswith(PYTHON_FILE_EXTENSIONS):
+            return True
```
conditions can be combined here.  Maybe make new constant CPP_AND_PYTHON_FILE_EXTENSIONS

jpeletier replied:
Done

bdraco commented on script/cpp_unit_test.py:
```diff
@@ -0,0 +1,206 @@
+LIST_COMPONENTS_SCRIPT = CURRENT_FILE.parent / "list-components.py"
+
+
+def get_all_components():
```
Please type everything in new python code

jpeletier replied:
Done

jpeletier committed:
cpp components test framework

linter

clang-tidy

improve CI speed

correct README

CI-friendly cpp unit testing

linting

linter2

add --all switch to cpp_unit_test

add unit testing to ci scripts

modify actions

ci.yml

lint

lint

Detect changes in cpp code

lint

type new python

reuse constants

lint

add a namespace to tests


pin ci actions

fix Path

fix test_determine_jobs

update clang-tidy hash

lint

jpeletier committed:
fix path

bdraco commented on script/determine-jobs.py:
```diff
@@ -103,6 +106,27 @@ class Platform(StrEnum):
+        cmd.extend(["-b", branch])
+
+    result = subprocess.run(cmd, capture_output=True, text=True, check=True)
+    return parse_list_components_output(result.stdout)
```
looks like some of this was moved to helpers

bdraco commented on script/list-components.py:
```diff
@@ -15,6 +23,100 @@ def get_all_component_files() -> list[str]:
+
+        return sorted(all_changed_components)
+
+    return sorted(components)
```
I think this was moved to helpers

bdraco commented on script/list-components.py:
```diff
@@ -15,6 +23,100 @@ def get_all_component_files() -> list[str]:
+                find_children_of_component(components_graph, child, depth + 1)
+            )
+    # Remove duplicate values
+    return list(set(children))
```
I think this was moved to helpers

jpeletier replied:
fixed

bdraco commented on script/list-components.py:
```diff
@@ -15,6 +23,100 @@ def get_all_component_files() -> list[str]:
+
+    if inspect.signature(auto_load).parameters:
+        return auto_load(config)
+    return auto_load()
```
I think this was moved to helpers

bdraco commented on script/list-components.py:
```diff
@@ -15,6 +23,100 @@ def get_all_component_files() -> list[str]:
+            components_graph[parent] = []
+        if child not in components_graph[parent]:
+            components_graph[parent].append(child)
+
```
I think this was moved to helpers

jpeletier committed:
remove redundant code after move to helpers.py

jpeletier committed:
document

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco commented on .github/workflows/ci.yml:
```diff
@@ -247,6 +249,45 @@ jobs:
+        uses: actions/cache@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
+        with:
+          path: venv
+          key: ${{ runner.os }}-${{ steps.python.outputs.python-version }}-venv-${{ needs.common.outputs.cache-key }}
```
need to check that this matches the other jobs and doesn't create conflicting caches

bdraco replied:
I think we have a helper for this

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
explict

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
dry

bdraco commented on script/determine-jobs.py:
```diff
@@ -283,6 +285,38 @@ def should_run_python_linters(branch: str | None = None) -> bool:
+        should run.
+    """
+    files = changed_files(branch)
+    if core_changed(files):
+        return ["--all"]
+
```
this is messy that we are mixing options and values

bdraco replied:
refactored it

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
Merge branch 'dev' into cpp-testing

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
Merge branch 'cpp-testing' of https://github.com/epiclabs-uc/esphome into cpp-testing

bdraco committed:
touch ups

bdraco commented on tests/components/uart/common.h:
```diff
@@ -0,0 +1,35 @@
+  using UARTComponent::write_array;
+  using UARTComponent::write_byte;
+
+  std::vector<uint8_t> written_data;
```
This is a bit ironic. We have been trying to get rid of temp vectors in the codebase. 

We should clean it up before merging, as it's very clear that people will start copying these right away

jpeletier replied:
fixed

bdraco committed:
comment

jpeletier committed:
removed temp vector

jpeletier committed:
Merge remote-tracking branch 'epic/cpp-testing' into cpp-testing

bdraco committed:
Merge branch 'dev' into cpp-testing
